### PR TITLE
Choice

### PR DESCRIPTION
**Pull request by jxv**
This is really a review request, not a PR.

### CODE REVIEW

**Comment by jxv** [comment]
It's worth reiterating that both body code for the fn parse_state is the same.

File: src/combinator.rs [hunk 1]
```diff
@@ -57,6 +57,65 @@ impl <P: Parser> Iterator for Iter<P> {
...
+    where P: Parser + 'a;
+
+impl <'a, I, O, P> Parser for ChoiceSlice<'a, P>
+    where I: Stream
+        , P: Parser<Input=I, Output=O> + Clone +  'a {
+    type Input = I;
+    type Output = O;
+    fn parse_state(&mut self, input: State<I>) -> ParseResult<O, I> {
+        if self.0.len() == 0 { // err: slice empty of parsers
+            return Err(Consumed::Empty(ParseError { position: input.position.clone(), errors: vec![] }))
```
Comment (Marwes) [inline src/combinator.rs]: Use the ParseError::new function instead. Since the fields are all public it is possible to construct like this but I defined the new function in some parts to make sure that there is always an error (no empty errors vector).

In this case something like `Error::Message("parser choice used with no parsers")`

File: src/combinator.rs [hunk 2]
```diff
@@ -57,6 +57,65 @@ impl <P: Parser> Iterator for Iter<P> {
...
+        }
+        for i in range(0, self.0.len()) {
```
Comment (Marwes) [inline src/combinator.rs]: Use self.0.iter()

File: src/combinator.rs [hunk 3]
```diff
@@ -57,6 +57,65 @@ impl <P: Parser> Iterator for Iter<P> {
...
+            match self.0[i].clone().parse_state(input.clone()) {
+                Err(_) => {},
```
Comment (Marwes) [inline src/combinator.rs]: You actually need to distinguish between `Err(Consumed(_))` and `Err(Empty(_))`. Check the `or` parser to see how it deals with this.

File: src/combinator.rs [hunk 4]
```diff
@@ -57,6 +57,65 @@ impl <P: Parser> Iterator for Iter<P> {
...
+                ok => return ok,
+            }
+        }
+        // err: no parsers matched
+        Err(Consumed::Empty(ParseError { position: input.position.clone(), errors: vec![] }))
```
Comment (Marwes) [inline src/combinator.rs]: Returning a new error here discards any information about the why the parsers didn't match.

File: src/combinator.rs [hunk 5]
```diff
@@ -57,6 +57,65 @@ impl <P: Parser> Iterator for Iter<P> {
...
+pub struct ChoiceVec<P>(Vec<P>)
+    where P: Parser;
+
+/* The body of parse_state is exactly the same text as ChoiceSlice's parse_state */
+impl <I, O, P> Parser for ChoiceVec<P>
+    where I: Stream
+        , P: Parser<Input=I, Output=O> + Clone {
+    type Input = I;
+    type Output = O;
+    fn parse_state(&mut self, input: State<I>) -> ParseResult<O, I> {
```
Comment (Marwes) [inline src/combinator.rs]: Replace entire function with `choice_slice(&*self.0).parse_state(input)` :)

**Comment by Marwes** [comment]
Added a few comments on your PR. I'm still not sold on the idea of including a choice parser though. While it may be useful in parsec, I think that being forced to box every parser makes it much less useful. If you or someone else can show a compelling use case for using choice over doing `p1.or(p2).or(p3)` I would be inclined to include it though.

**Comment by Marwes** [comment]
Instead of having choice_slice and choice_vec, maybe it is possible to only have
```rust
fn choice<I>(ps: I) -> Choice<I> where I: Iterator + Clone, Iterator::Item: Parser
```

**Comment by jxv** [comment]
Here's my case to use ChoiceSlice: https://github.com/jxv/irc/commit/5d4d82aa2362b5282db5d674b3de3ad3565f461c#diff-9c0b8390b3d65627e2d9c5ad4536cc35R242
Scroll down a bit to `let command = ...`

I haven't actually got it to compile using the `or` equivalent because it takes a *really* long time. After an hour, I'd Ctrl-D it. It works with smaller `or`'s though.

**Comment by jxv** [comment]
This gives crashes the compiler:
```rust
fn choice<I>(ps: I) -> Choice<I> where I: Iterator + Clone, Iterator::Item: Parser
```
Either I'm doing something wrong or there's a legit bug.

**Comment by Marwes** [comment]
Your problem with the compile times is this issue https://github.com/rust-lang/rust/issues/21231. Specializing the `or` parser before using could work for you but I can't be certain of course. That does seem like a common enough use case though to warrant the choice parsers inclusion.

I noticed you have a bug though in the parser, since it is LL(1) parsers that are created you need to use the `try` parser to get more look ahead. (Surround each `pc::string` parser call with `try`).
Also you can use the `map` function instead of `parser.with(value(...)))` which I find slightly clearer

**Comment by Marwes** [comment]
Anything that crashes the compiler is a bug and the code is actually wrong. It should be:

```
fn choice<I>(ps: I) -> Choice<I> where I: Iterator + Clone, <I as Iterator>::Item: Parser
```

**Comment by Marwes** [comment]
This seems to be the issue in rust #21867

**Comment by jxv** [comment]
So, okay. It's bug (or two).

Also, thanks for the `pc::try` tip! Although I don't think the `parser.with(value(x))` matters much, it feels clear to me because it somewhat resembles `parser >> return x`.

Other than the compiler bug with the choice traits, is there anything else which can be improved?

**Comment by Marwes** [comment]
A couple of things left I think.

I'd argue that ChoiceSlice should hold `&mut [P]` which would allow you to remove the Clone bound on P in all places.

I think you got the ChoiceSlice right though it was a little hard to follow the code. I think it may be easier to visualize if you think of the parser as returning three different values you need to deal with.

* Ok(_) => means we are done, just return it (which you have done)
* Err(Consumed(_)) => Since it is a LL(1) parser there is nothing more  to be done here either, just return the error
* Err(Empty(_)) => This is the tricky one. If there is no error so far the error needs to be saved, if an Empty error has been saved since later the old and new error needs to be merged together.

I think you got all of these right but a test or two for the parser would be really useful. Only thing that is 'wrong' is that you don't exit early when a consumed error occurs (which is probably why you didn't notice that you needed try in your parsers). Arguably you should only construct the "choice is empty" error if the slice is actually empty, it should avoid an unnecessary allocation.

This ended up a longer response than I expected but choice is probably one of the trickier parsers to get right.

I should probably add some function which helps deal with these cases correctly though, since I know I had to write some extra boilerplate similar to this for a few other parsers (there is the `combine` method on `Consumed` but it doesn't work for this case).

**Comment by jxv** [comment]
Yeah, I think my misunderstanding of the relationship between ```try``` and ```Consumed``` attributed to the complexity. Thank you for explaning it thoroughly.

In regards to the ```&mut [P]``` and ```Clone``` for ChoiceSlice, how about using ```Rc``` instead?

**Comment by Marwes** [comment]
How would you use `Rc`? I don't really see it since if you have a parser inside an `Rc` you wouldn't be able to get a mutable reference to it.

I have experimented a bit with having parsers take themselves as &self but I didn't feel like it was a good idea since it would force all parsers to take functions as `Fn` instead of `FnMut`.

**Comment by jxv** [comment]
I was thinking that ref-counting might have been reduced the overhead from cloning. But yeah, you'd lose the ability to mutate.

**Comment by jxv** [comment]
Is this test okay?  https://github.com/Marwes/parser-combinators/pull/16/files#diff-b4aea3e418ccdb71239b96952d9cddb6R395

**Comment by Marwes** [comment]
Looks fine to me, haven't done many tests which are more thorough than that. It might be good to have one call to parse which returns `Err` but I'd say it is good enough regardless.

**Comment by Marwes** [comment]
If you feel like you have nothing more to add to this I can merge this after a rebase (I can do the rebase if you want to, just remember to pull in the new master afterwards and work from that).

**Comment by jxv** [comment]
Yes, you can rebase and merge.

**Comment by Marwes** [comment]
Rebased and merged in 3bc873941a37834cc45f9d33981bfef3b1407028
